File: exercises/ddd_infrastructure_layer/typescript/test/application/services/RejectService.test.ts
import { describe, expect, it } from "bun:test";
import { RejectService } from "../../../src/application/services/RejectService";
import { RecordingApplicationRepository } from "../../helpers/RecordingApplicationRepository";
import { ApplicationStatus } from "../../../src/domain/application/ApplicationStatus";
import { buildApplicationWithMotivationLetter } from "../../helpers/applicationFixtures";

const applicationId = "33333333-3333-3333-3333-333333333333";

describe("RejectService", () => {
  it("throws when no application exists", async () => {
    const repo = new RecordingApplicationRepository();
    const service = new RejectService(repo);

    await expect(
      service.execute({ applicationId })
    ).rejects.toThrow(`Application not found for id ${applicationId}`);
  });

  it("marks the application as rejected and publishes an event", async () => {
    const existing = buildApplicationWithMotivationLetter(applicationId);
    const repo = new RecordingApplicationRepository([existing]);
    const service = new RejectService(repo);

    await service.execute({ applicationId });

    expect(repo.saved.length).toBe(1);
    const saved = repo.saved[0];
    expect(saved!.status).toBe(ApplicationStatus.Rejected);
  });
});

File: exercises/ddd_infrastructure_layer/java/src/test/java/application/services/RejectServiceTest.java
package application.services;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.UUID;

import org.junit.jupiter.api.Test;

import application.services.RejectService.RejectCommand;
import domain.application.ApplicationStatus;
import helpers.ApplicationFixtures;
import helpers.RecordingApplicationRepository;

public class RejectServiceTest {
    private static final UUID APPLICATION_ID = UUID.fromString("55555555-5555-5555-5555-555555555555");

    @Test
    void testThrowsWhenNoApplicationCanBeFound() {
        var repo = new RecordingApplicationRepository();
        var service = new RejectService(repo);

        assertThrows(IllegalStateException.class, () -> {
            service.execute(new RejectCommand(APPLICATION_ID));
        });
    }

    @Test
    void testMarksApplicationAsRejected() throws Exception {
        var existing = ApplicationFixtures.buildApplicationWithMotivationLetter(APPLICATION_ID);
        var repo = new RecordingApplicationRepository(existing);
        var service = new RejectService(repo);

        service.execute(new RejectCommand(APPLICATION_ID));

        var saved = repo.saved.get(0);
        assertEquals(ApplicationStatus.REJECTED, saved.getStatus());
    }
}
